sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,284 @@
+    >>> import numpy as np
+
+    >>> X = np.asarray([[1, 2], [1, 4], [1, 0],
+    ...                 [4, 2], [4, 4], [4, 0]])
+    >>> kmedoids = KMedoids(n_clusters=2, random_state=0).fit(X)
+    >>> kmedoids.labels_
+    array([0, 0, 0, 1, 1, 1])
+    >>> kmedoids.predict([[0,0], [4,4]])
+    array([0, 1])
+    >>> kmedoids.cluster_centers_
+    array([[1, 2],
+           [4, 2]])
+    >>> kmedoids.inertia_
+    8.0
+
+    >>> kmedoids = KMedoids(n_clusters=2, random_state=0,
+    ...                     distance_metric='manhattan').fit(X)
+    >>> kmedoids.labels_
+    array([0, 0, 0, 1, 1, 1])
+    >>> kmedoids.predict([[0,0], [4,4]])
+    array([0, 1])
+    >>> kmedoids.cluster_centers_
+    array([[1, 2],
+           [4, 2]])
+    >>> kmedoids.inertia_
+    8.0
+
+    References
+    ----------
+    Kaufman, L. and Rousseeuw, P.J., Statistical Data Analysis Based on
+    the L1–Norm and Related Methods, edited by Y. Dodge, North-Holland,
+    405–416. 1987
+    """
+
+    def __init__(self, n_clusters=8, distance_metric='euclidean',
+                 init='heuristic', max_iter=300, random_state=None):
+        self.n_clusters = n_clusters
+        self.distance_metric = distance_metric
+        self.init = init
+        self.max_iter = max_iter
+        self.random_state = random_state
+
+    def _check_init_args(self):
+        """Validates the input arguments. """
+
+        # Check n_clusters
+        if (self.n_clusters is None or self.n_clusters <= 0 or
+                not isinstance(self.n_clusters, (int, np.integer))):
+            raise ValueError("n_clusters should be a nonnegative integer. "
+                             "%s was given" % self.n_clusters)
+
+        # Check init
+        init_methods = ['random', 'heuristic']
+        if self.init not in init_methods:
+            raise ValueError("init needs to be one of " +
+                             "the following: " +
+                             "%s" % init_methods)
+
+        # Check random state
+        self.random_state_ = check_random_state(self.random_state)
+
+    def fit(self, X, y=None):
+        """Fit K-Medoids to the provided data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features).
+            Dataset to cluster.
+
+        Returns
+        -------
+        self
+        """
+
+        self._check_init_args()
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+        if self.n_clusters > X.shape[0]:
+            raise ValueError("The number of medoids %d must be less "
+                             "than the number of samples %d."
+                             % (self.n_clusters, X.shape[0]))
+
+        distances = pairwise_distances(X, metric=self.distance_metric)
+
+        medoid_idxs = self._get_initial_medoid_indices(distances,
+                                                       self.n_clusters)
+        labels = np.argmin(distances[medoid_idxs, :], axis=0)
```

Comment:
This is never used unless max_iter=0.

Reply:
That's true. Should I add a check to _check_init_args for max_iter > 0 ?

Reply:
I suppose so

Reply:
done